system/datatypes/settings.php

```diff
@@ -327,6 +342,109 @@ function saveTheme($argument) {
 			return 'reload';
 		}
 
+		function copyTheme($argument) {
+			if (isAdmin()) {
+				$errors = $this->validateAndCopyTheme();
+				foreach ($errors as $error) notify('error', $error);
+				if (empty($errors)) {
+					notify('success', __('copied-theme-successful'));
+				}
+			}
+			return 'reload';
+		}
+
+		function copyThemeAndActivate($argument) {
+			if (isAdmin()) {
+				$errors = $this->validateAndCopyTheme();
+				foreach ($errors as $error) notify('error', $error);
+				if (empty($errors)) {
+					global $hyphaXml;
+					$hyphaXml->lockAndReload();
+					hypha_setTheme($this->toSafeDir($_POST['dstTheme']));
+					$hyphaXml->saveAndUnlock();
+					notify('success', __('copied-theme-successful'));
+				}
+			}
+			return 'reload';
+		}
+
+		function validateAndCopyTheme() {
+			// get list of existing themes
+			$themes = $this->getThemes();
+
+			$errors = [];
+			// validate dstTheme
+			$dstTheme = isset($_POST['dstTheme']) && '' != $_POST['dstTheme'] ? $this->toSafeDir($_POST['dstTheme']) : null;
+			if (null === $dstTheme) {
+				$errors[] = __('destination-theme-name-required');
+			} elseif (in_array($dstTheme, $themes)) {
+				$errors[] = __('destination-theme-name-already-taken');
+			}
+
+			// validate srcTheme
+			if (!isset($_POST['srcTheme']) || '' == $_POST['srcTheme']) {
+				$errors[] = __('source-theme-name-required');
+			} elseif (!in_array($_POST['srcTheme'], $themes)) {
+				$errors[] = __('source-theme-name-not-found');
+			}
+
+			// return the errors if there are any
+			if (empty($errors)) {
+				// create an anonymous function to cursively copy the theme directory
+				$dirCopy = function ($src, $dst) use (&$dirCopy) {
+					mkdir($dst, 0744);
+					foreach (scandir($src) as $file) {
+						if (in_array($file, ['.', '..'])) {
+							continue;
+						}
+						if (is_dir($src . '/' . $file)) {
+							$dirCopy($src . '/' . $file, $dst . '/' . $file);
+						} else {
+							copy($src . '/' . $file, $dst . '/' . $file);
+						}
+					}
+				};
+				$dirCopy('data/themes/' . $_POST['srcTheme'], 'data/themes/' . $dstTheme);
+			}
+
+			return $errors;
+		}
+
+		/**
+		 * Convert given directory name into a safe directory name for url and file system.
+		 *
+		 * @param string $dir
+		 * @return string
+		 */
+		function toSafeDir($dir) {
+			// Strip HTML Tags
+			$dir = strip_tags($dir);
+			// Clean up things like &amp;
+			$dir = html_entity_decode($dir);
+			// Strip out any url-encoded stuff
+			$dir = urldecode($dir);
+			// Replace non-AlNum characters with space
+			$dir = preg_replace('/[^A-Za-z0-9]/', ' ', $dir);
+			// Replace Multiple spaces with single space
+			$dir = preg_replace('/ +/', ' ', $dir);
+			// Trim the string of leading/trailing space
+			$dir = trim($dir);
+			// Replace spaces with dashes
+			$dir = preg_replace('/ /', '-', $dir);
```

Comment:
Why do the first three things? There is no reason for any of these encodings to be present AFAICS, so they could only be there when a user types them manually. In that case, the next regex (non-alnum) will strip away such encodings anyway to make things harmless. In fact, I think *only* the alnum regex could actually be sufficient already (it would be safe), though the operations after that could make a string with non alnum stuff nicer to read (but it might still be unexpected for a user that his carefully typed string is chopped up). Perhaps it is nicer to do this checking client-side already (either flag invalid values or prevent typing invalid chars in the first place). I suspect that HTML5 might have something convenient for this? Of course, a server side check/replace is still needed, but that can then just be the alnum regex above, no need to generate something pretty, only safe.

We could consider allowing both - and _ in names as well? . is also obvious, but opens the door to path traversal (though maybe not without /).

Reply:
This is a copy-paste from https://stackoverflow.com/a/7128879, I didn't want to spend too much time digging into this.

Reply:
Solved in an other way.